core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -170,89 +189,243 @@ public boolean inodeIdExists(long id) {
   }
 
   /**
-   * Returns a list of existing inodes on the given path.
+   * Attempts to extend an existing {@link LockedInodePath} to reach the target inode (the last
+   * inode for the full path). If the target inode does not exist, an exception will be thrown.
    *
-   * @param path the path to get the inodes list for
-   * @return the inodes list with the given path
+   * @param inodePath the {@link LockedInodePath} to extend to the target inode
+   * @param lockMode the {@link LockMode} to lock the inodes with
    * @throws InvalidPathException if the path is invalid
+   * @throws FileDoesNotExistException if the target inode does not exist
    */
-  public List<Inode<?>> collectInodes(AlluxioURI path) throws InvalidPathException {
-    TraversalResult traversalResult =
-        traverseToInode(PathUtils.getPathComponents(path.getPath()), false);
-    return traversalResult.getInodes();
+  public void ensureFullInodePath(LockedInodePath inodePath, LockMode lockMode)
+      throws InvalidPathException, FileDoesNotExistException {
+    TraversalResult traversalResult = traverseToInode(inodePath, lockMode);
+    if (!traversalResult.isFound()) {
+      throw new FileDoesNotExistException(
+          ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(inodePath.getUri()));
+    }
+  }
+
+  private void computePathForInode(Inode<?> inode, StringBuilder builder)
+      throws FileDoesNotExistException {
+    inode.lockRead();
+    long id = inode.getId();
+    long parentId = inode.getParentId();
+    String name = inode.getName();
+    inode.unlockRead();
+
+    if (isRootId(id)) {
+      builder.append(AlluxioURI.SEPARATOR);
+    } else if (isRootId(parentId)) {
+      builder.append(AlluxioURI.SEPARATOR);
+      builder.append(name);
+    } else {
+      Inode<?> parentInode = mInodes.getFirstByField(mIdIndex, parentId);
+      if (parentInode == null) {
+        throw new FileDoesNotExistException("Inode id " + parentId + " does not exist.");
+      }
+
+      computePathForInode(parentInode, builder);
+      builder.append(AlluxioURI.SEPARATOR);
+      builder.append(name);
+    }
   }
 
   /**
+   * Returns the path for a particular inode. The inode and the path to the inode must already be
+   * locked.
+   *
    * @param inode the inode to get the path for
-   * @return the path for a given inode
+   * @return the {@link AlluxioURI} for the path of the inode
+   * @throws FileDoesNotExistException if the path does not exist
    */
-  public AlluxioURI getPath(Inode<?> inode) {
-    if (isRootId(inode.getId())) {
-      return new AlluxioURI(AlluxioURI.SEPARATOR);
-    }
-    if (isRootId(inode.getParentId())) {
-      return new AlluxioURI(AlluxioURI.SEPARATOR + inode.getName());
-    }
-    return getPath(mInodes.getFirstByField(mIdIndex, inode.getParentId())).join(inode.getName());
+  public AlluxioURI getPath(Inode<?> inode) throws FileDoesNotExistException {
+    StringBuilder builder = new StringBuilder();
```

Comment:
consider verifying the caller indeed holds the readlock by exposing an inode method that ends up calling `getReadHoldCount()`